src/java.net.http/share/classes/jdk/internal/net/http/Http1Request.java

```diff
@@ -298,14 +298,20 @@ synchronized void setFinished() {
             contentLength = requestPublisher.contentLength();
         }
 
-        if (contentLength == 0) {
-            systemHeadersBuilder.setHeader("Content-Length", "0");
-        } else if (contentLength > 0) {
-            systemHeadersBuilder.setHeader("Content-Length", Long.toString(contentLength));
-            streaming = false;
-        } else {
-            streaming = true;
-            systemHeadersBuilder.setHeader("Transfer-encoding", "chunked");
+        // GET with no body should not set the Content-Length header
+        if (requestPublisher != null || !"GET".equals(request.method())) {
```

Comment:
Can we remove the check for "GET"? This way we will let the users decide if they want to send content-length or not, regardless of the chosen request method.

Reply:
Practically that would mean not sending Content-Length: 0 by default for GET, DELETE, and HEAD. All other requests methods would have either a Content-Length or Transfer-Encoding. I suspect that HEAD should probably be handled the same way than GET. But should DELETE not have a body? I know that some servers will balk if DELETE has a body. But would they expect Content-Length: 0 not to be included? @Michael-Mc-Mahon what do you think?

Reply:
RFC 7230 section 3.3.2 says:
"A user agent SHOULD NOT send a
   Content-Length header field when the request message does not contain
   a payload body and the method semantics do not anticipate such a
   body."
I don't think the DELETE method anticipates a a request body. So, it shouldn't have a Content-Length header either IMO. Same for HEAD also.

Reply:
Thanks Michael. Let's do as @djelinski daniel suggests then - and only set Content-Length if requestPublisher != null (that is remove the || !"GET".equals(request.method()) part of the if). The test would need to be extended to test DELETE and HEAD but that shouldn't be too difficult.

Reply:
FWIW, the relevant text now is in https://www.rfc-editor.org/rfc/rfc9110.html#section-8.6:

> A user agent SHOULD send Content-Length in a request when the method defines a meaning for enclosed content and it is not sending Transfer-Encoding. For example, a user agent normally sends Content-Length in a POST request even when the value is 0 (indicating empty content). A user agent SHOULD NOT send a Content-Length header field when the request message does not contain content and the method semantics do not anticipate such data.